### PR TITLE
WIP: update to tko

### PR DESCRIPTION
**Pull request by piever**

### CODE REVIEW

File: src/Knockout.jl [hunk 1]
```diff
@@ -127,6 +126,6 @@ isnumeric(x::Number) = true
 isnumeric(x::Bool) = false
 isnumeric(x::AbstractObservable) = isnumeric(x[])
 
-js_lambda(s::String) = "function (){$s}"
+js_lambda(s::String) = "=> $s"
```
Comment (twavv) [inline src/Knockout.jl]: Shouldn't this be `"() => $s"`?

**Review by twavv — CHANGES_REQUESTED** [review]
I think the JS-lambda change you're proposing is breaking.

```julia
template = dom"button"(
    "{{message}}",
    attributes=Dict(
        # This is broken.
        "data-bind" => "click: () => this.message('goodbye!')",
        # This works though.
        # "data-bind" => "click: function(){ this.message('goodbye!') }",
    ),
)
message = Observable{String}("Hello!")
knockout(template, [:message=>message])
```

I'm 99% sure that the reason is because arrow functions can't be bound in JavaScript (the whole reason they were introduced was to have 100% certainty that the value of `this` was correct, regardless of how the method was invoked). So when you use an arrow function, `this` doesn't refer to the correct Knockout object, it refers to something else since knockout wasn't able to bind the correct thing to `this`.

File: src/Knockout.jl (hunk 1 above)
Comment (piever) [reply src/Knockout.jl]: Yes, I guess that'd be cleaner!

**Comment by piever** [comment]
Actually, what happened is that Knockout is being updated to Knockout 4.0 and they switched from using anonymous functions to lambdas (there is some discussion here https://github.com/knockout/tko/issues/91). I confess I don't have a clear understanding of the lambda versus anonymous function difference, for more complicated things I think it is recommended to pass them to the Knockout model as `methods`.

If you build Knockout on this PR, it should download the correct Knockout javascript (4.0 alpha 8) and your snippet above should work. I still have not made up my mind whethere there is a good reason to update or not (the new codebase should be cleaner, smaller and more efficient but may be less tested than the old one).

**Comment by piever** [comment]
The main drawback seems to be that nesting widgets stops working (we had the same problem with Vue.js which is why we moved to Knockout), for example:

```julia
julia> s = tabulator(OrderedDict("a" => checkbox("Asda"), "b" => slider(1:100)));
```

gives `Error: You cannot apply bindings multiple times to the same element`, so for sure better defer this till the nesting problem gets clarified. If you have a simple way to translate this problem into a pure javascript + HTML MWE it may be useful to report to the Technical Knockout repo.

**Review by twavv — APPROVED** [review]

**Comment by twavv** [comment]
I remember getting that issue before when writing the new client side of WebIO. I think it had something to do with the order in which elements were constructed versus when scripts were run. It also might be possible that the way nested things work currently is just an implementation of Knockout? Do they support nested things?

**Comment by piever** [comment]
I think they support that, I'm asking there how I can prevent a node from getting bindings from a model in a parent node (somehow that seemed to work out of the box before).

**Comment by piever** [comment]
It seems that you can add this to the Knockout model and then you'd need to set `data-bind="allowBindings: false` in a node if you don't want the parent model to be applied there. ~Could we maybe even set this by default in all scopes from WebIO?~ On a second thought maybe that's too much, it should be up to the widget that uses knockout and has children nodes to decide whether they should also get the bindings or not.

**Comment by twavv** [comment]
I don't think that should happen in WebIO itself (at least in an ideal world) since it's specific to Knockout. Could you just wrap knockouts in a node that has that attribute?

**Pull request closed without merging**